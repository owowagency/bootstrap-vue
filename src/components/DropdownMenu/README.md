# DropdownMenu

```vue
<DropdownMenu :items="[Item]" />
```

## Imported Props

| Prop name | Description                    | Type                                         | Values | Default  |
| --------- | ------------------------------ | -------------------------------------------- | ------ | -------- |
| items     | The items of the dropdown menu | [Item[]](../../composables/useDropdownItems) | -      | () => [] |

## Slots

| Name    | Description                               | Bindings                                                                                                           |
| ------- | ----------------------------------------- | ------------------------------------------------------------------------------------------------------------------ |
| prepend | Prepends the dropdown menu                |                                                                                                                    |
| items   | Displays content inside the dropdown menu | **click-item** `function` - The function to call when an item is clicked<br/>**items** `array` - The `Item` list   |
| item    | Displays the dropdown menu item           | **click-item** `function` - The function to call when an item is clicked<br/>**item** `object` - The `Item` object |
| append  | Appends the dropdown menu                 |                                                                                                                    |

## Events

| Event name | Properties                                  | Description                   |
| ---------- | ------------------------------------------- | ----------------------------- |
| click:item | **item** `Item` - The item that was clicked | Fired when an item is clicked |
