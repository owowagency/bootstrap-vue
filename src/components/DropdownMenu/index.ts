import DropdownMenu from './DropdownMenu.vue';

export default DropdownMenu;
