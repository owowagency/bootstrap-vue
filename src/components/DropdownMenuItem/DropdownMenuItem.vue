<template>
    <li>
        <hr
            v-if="divider"
            class="dropdown-divider"
            :class="itemClass"
        >

        <h6
            v-else-if="header"
            class="dropdown-header"
            :class="itemClass"
        >
            <!-- @slot Displays content inside the dropdown menu item. -->
            <slot name="default">
                {{ label }}
            </slot>
        </h6>

        <Component
            :is="is"
            v-else
            class="dropdown-item"
            :class="[{active, disabled}, itemClass]"
            :href="href"
            :to="to"
            @click.prevent
        >
            <slot name="default">
                {{ label }}
            </slot>
        </Component>
    </li>
</template>

<script lang="ts" setup>
import {PropType, computed} from 'vue';
import {dropdownItemProps} from '@/composables/useDropdownItem';

const props = defineProps({
    /**
     * Adds the `active` class
     */
    active: {
        type: Boolean,
        default: false,
    },
    /**
     * Adds the `disabled` class
     */
    disabled: {
        type: Boolean,
        default: false,
    },
    /**
     * Displays a divider
     */
    divider: {
        type: Boolean,
        default: false,
    },
    ...dropdownItemProps,
    /**
     * Displays label as header
     */
    header: {
        type: Boolean,
        default: false,
    },
    href: {
        type: String,
        default: '#',
    },
    is: {
        type: String,
        default: undefined,
    },
    /**
     * The label to display
     */
    label: {
        type: String,
        default: undefined,
    },
    to: {
        type: [Object, String] as PropType<string|Record<string, unknown>>,
        default: undefined,
    },
});

const is = computed(() => {
    if (props.is) {
        return props.is;
    }

    if (props.to) {
        return 'RouterLink';
    }

    return 'a';
});
</script>

<docs>
```vue
<DropdownMenuItem label="Menu item" />
```
</docs>
