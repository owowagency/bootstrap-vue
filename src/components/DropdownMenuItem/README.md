# DropdownMenuItem

```vue
<DropdownMenuItem label="Menu item" />
```

## Props

| Prop name | Description               | Type                                  | Values | Default   |
| --------- | ------------------------- | ------------------------------------- | ------ | --------- |
| active    | Adds the `active` class   | boolean                               | -      | false     |
| disabled  | Adds the `disabled` class | boolean                               | -      | false     |
| divider   | Displays a divider        | boolean                               | -      | false     |
| header    | Displays label as header  | boolean                               | -      | false     |
| href      |                           | string                                | -      | '#'       |
| is        |                           | string                                | -      | undefined |
| label     | The label to display      | string                                | -      | undefined |
| to        |                           | string\|Record&lt;string, unknown&gt; | -      | undefined |

## Slots

| Name    | Description                                     | Bindings |
| ------- | ----------------------------------------------- | -------- |
| default | Displays content inside the dropdown menu item. |          |
