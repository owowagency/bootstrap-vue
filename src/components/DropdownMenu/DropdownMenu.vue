<template>
    <ul class="dropdown-menu">
        <!-- @slot Prepends the dropdown menu -->
        <slot name="prepend" />

        <!--
        @slot Displays content inside the dropdown menu
        @binding {array} items The `Item` list
        @binding {function} click-item The function to call when an item is clicked
        -->
        <slot
            name="items"
            :click-item="clickItem"
            :items="items"
        >
            <template
                v-for="(item, index) of items"
                :key="`item-${index.toString()}`"
            >
                <!--
                @slot Displays the dropdown menu item
                @binding {object} item The `Item` object
                @binding {function} click-item The function to call when an item is clicked
                -->
                <slot
                    name="item"
                    :click-item="clickItem"
                    :item="item"
                >
                    <DropdownMenuItem
                        v-bind="item"
                        :item-class="itemClass"
                        @click="clickItem(item)"
                    />
                </slot>
            </template>
        </slot>

        <!-- @slot Appends the dropdown menu -->
        <slot name="append" />
    </ul>
</template>

<script lang="ts">
import {Item, dropdownItemsProps} from '@/composables/useDropdownItems';
import DropdownMenuItem from '@/components/DropdownMenuItem';
import {dropdownItemProps} from '@/composables/useDropdownItem';
</script>

<script lang="ts" setup>
defineProps({
    ...dropdownItemProps,
    ...dropdownItemsProps,
});

const emit = defineEmits([
    /**
     * Fired when an item is clicked
     * @param {Item} item The item that was clicked
     */
    'click:item',
]);

const clickItem = (item: Item) => emit('click:item', item);
</script>

<docs>
```vue
<DropdownMenu :items="[Item]" />
```

## Imported Props

| Prop name | Description                    | Type                                         | Values | Default  |
| --------- | ------------------------------ | -------------------------------------------- | ------ | -------- |
| items     | The items of the dropdown menu | [Item[]](../../composables/useDropdownItems) | -      | () => [] |
</docs>
