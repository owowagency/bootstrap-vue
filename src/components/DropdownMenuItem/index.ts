import DropdownMenuItem from './DropdownMenuItem.vue';

export default DropdownMenuItem;
