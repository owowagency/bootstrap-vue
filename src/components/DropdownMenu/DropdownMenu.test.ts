import DropdownMenu from '@/components/DropdownMenu';
import DropdownMenuItem from '@/components/DropdownMenuItem';
import {shallowMount} from '@vue/test-utils';

const items = [
    {header: true, label: 'Item 1'},
    {label: 'Item 3'},
    {divider: true},
    {label: 'Item 4', active: true},
];

describe('template', () => {
    componentRenderTest(DropdownMenu, {props: {items}});

    componentSlotRenderTest(DropdownMenu, 'prepend');

    componentSlotRenderTest(DropdownMenu, 'items');

    componentSlotRenderTest(DropdownMenu, 'item', {props: {items}});

    componentSlotRenderTest(DropdownMenu, 'append');

    it('emits event on click dropdown menu item', async() => {
        const wrapper = shallowMount(DropdownMenu, {props: {items}});

        await wrapper.findComponent(DropdownMenuItem).trigger('click');

        expect(wrapper.emitted('click:item')[0]).toEqual([items[0]]);
    });
});
