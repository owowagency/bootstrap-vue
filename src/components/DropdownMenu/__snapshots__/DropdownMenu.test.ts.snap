// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<ul class=\\"dropdown-menu\\">
  <!-- @slot Prepends the dropdown menu -->
  <!--
        @slot Displays content inside the dropdown menu
        @binding {array} items The \`Item\` list
        @binding {function} click-item The function to call when an item is clicked
        -->
  <!--
                @slot Displays the dropdown menu item
                @binding {object} item The \`Item\` object
                @binding {function} click-item The function to call when an item is clicked
                -->
  <dropdown-menu-item-stub active=\\"false\\" disabled=\\"false\\" divider=\\"false\\" header=\\"true\\" href=\\"#\\" label=\\"Item 1\\"></dropdown-menu-item-stub>
  <!--
                @slot Displays the dropdown menu item
                @binding {object} item The \`Item\` object
                @binding {function} click-item The function to call when an item is clicked
                -->
  <dropdown-menu-item-stub active=\\"false\\" disabled=\\"false\\" divider=\\"false\\" header=\\"false\\" href=\\"#\\" label=\\"Item 3\\"></dropdown-menu-item-stub>
  <!--
                @slot Displays the dropdown menu item
                @binding {object} item The \`Item\` object
                @binding {function} click-item The function to call when an item is clicked
                -->
  <dropdown-menu-item-stub active=\\"false\\" disabled=\\"false\\" divider=\\"true\\" header=\\"false\\" href=\\"#\\"></dropdown-menu-item-stub>
  <!--
                @slot Displays the dropdown menu item
                @binding {object} item The \`Item\` object
                @binding {function} click-item The function to call when an item is clicked
                -->
  <dropdown-menu-item-stub active=\\"true\\" disabled=\\"false\\" divider=\\"false\\" header=\\"false\\" href=\\"#\\" label=\\"Item 4\\"></dropdown-menu-item-stub><!-- @slot Appends the dropdown menu -->
</ul>"
`;
